interpreter/valid/valid.ml

```diff
@@ -496,64 +496,38 @@ let rec check_instr (c : context) (e : instr) (s : infer_result_type) : infer_in
     [RefT (NoNull, DefHT y)] -->
     [RefT (NoNull, DefHT (Stat x.it))], []
 
-  | ContBind x ->
-    (match peek_ref 0 s e.at with
-    | nul, DefHT (Stat y) ->
-      let ContT z = cont_type c (y @@ e.at) in
-      let FuncT (ts1, ts2) = func_type c (as_stat_var z @@ e.at) in
-      let ContT z' = cont_type c x in
-      let FuncT (ts1', _) as ft' = func_type c (as_stat_var z' @@ x.at) in
-      require (List.length ts1 >= List.length ts1') x.at
-        "type mismatch in continuation arguments";
-      let ts11, ts12 = Lib.List.split (List.length ts1 - List.length ts1') ts1 in
-      require (match_func_type c.types (FuncT (ts12, ts2)) ft') e.at
-        "type mismatch in continuation type";
-      (ts11 @ [RefT (nul, DefHT (Stat y))]) -->
-        [RefT (NoNull, DefHT (Stat x.it))], []
-    | (_, BotHT) as rt ->
-      [RefT rt] -->.. [RefT (NoNull, DefHT (Stat x.it))], []
-    | rt ->
-      error e.at
-        ("type mismatch: instruction requires continuation reference type" ^
-         " but stack has " ^ string_of_val_type (RefT rt))
-    )
+  | ContBind (x, y) ->
+    let (nul, _) = peek_ref 0 s e.at in
```

Comment:
With the type annotation you  shouldn't need the peek anymore, but simply use Null as the principal type.

Reply:
OK!